src/vos/vos_dtx.c

```diff
@@ -1101,3 +1627,263 @@ vos_dtx_mark_sync(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch)
+		if (bcb == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+
+		bcb->bcb_dae_count = dsb->dsb_count;
+		if (bcb->bcb_dae_count == 0) {
+			if (cont_df->cd_dtx_active_tail != dsb_off ||
+			    dsb->dsb_index >= dsb->dsb_cap) {
+				D_ERROR("Invalid active DTX blob: "UMOFF_PF
+					", idx %d, cap %d\n", UMOFF_P(dsb_off),
+					dsb->dsb_index, dsb->dsb_cap);
+				rc = -DER_IO;
+			}
+
+			D_GOTO(out, rc);
+		}
+
+		for (i = 0; i < dsb->dsb_index; i++) {
+			struct vos_dtx_act_ent_df	*dae_df;
+			struct vos_dtx_act_ent		*dae;
+			int				 count;
+
+			dae_df = &dsb->dsb_active_data[i];
+			rc = dtx_act_handle_batched_cleanup(cont, bcb, dae_df,
+							    &dsb_off);
+			if (rc > 1)
+				goto next;
+
+			if (rc == 1)
+				continue;
+
+			if (rc < 0)
+				D_GOTO(out, rc);
+
+			D_ALLOC_PTR(dae);
+			if (dae == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+
+			memcpy(&dae->dae_base, dae_df, sizeof(dae->dae_base));
+			dae->dae_df_off = umem_ptr2off(umm, dae_df);
+			dae->dae_dsb = dsb;
+			dae->dae_bcb = bcb;
+
+			if (DAE_REC_CNT(dae) <= DTX_INLINE_REC_CNT)
+				goto insert;
+
+			count = DAE_REC_CNT(dae) - DTX_INLINE_REC_CNT;
+			D_ALLOC(dae->dae_records,
+				sizeof(*dae->dae_records) * count);
+			if (dae->dae_records == NULL) {
+				D_FREE_PTR(dae);
+				D_GOTO(out, rc = -DER_NOMEM);
+			}
+
+			memcpy(dae->dae_records,
+			       umem_off2ptr(umm, dae_df->dae_rec_off),
+			       sizeof(*dae->dae_records) * count);
+			dae->dae_rec_cap = count;
+
+insert:
+			d_iov_set(&kiov, &DAE_XID(dae), sizeof(DAE_XID(dae)));
+			d_iov_set(&riov, dae, sizeof(*dae));
+			rc = dbtree_upsert(cont->vc_dtx_active_hdl,
+					   BTR_PROBE_EQ, DAOS_INTENT_UPDATE,
+					   &kiov, &riov);
+			if (rc != 0) {
+				D_FREE(dae->dae_records);
+				D_FREE_PTR(dae);
+				goto out;
+			}
+		}
+
+		dsb_off = dsb->dsb_next;
+	}
+
+out:
+	return rc > 0 ? 0 : rc;
+}
+
+int
+vos_dtx_cmt_reindex(daos_handle_t coh, void *hint)
+{
+	struct umem_instance		*umm;
+	struct vos_container		*cont;
+	struct vos_cont_df		*cont_df;
+	struct vos_dtx_cmt_ent		*dce;
+	struct vos_dtx_scm_blob		*dsb;
+	umem_off_t			*dsb_off = hint;
+	d_iov_t				 kiov;
+	d_iov_t				 riov;
+	int				 rc = 0;
+	int				 i;
+
+	cont = vos_hdl2cont(coh);
+	D_ASSERT(cont != NULL);
+
+	umm = vos_cont2umm(cont);
+	cont_df = cont->vc_cont_df;
+
+	if (dtx_is_null(*dsb_off))
+		dsb = umem_off2ptr(umm, cont_df->cd_dtx_committed_head);
+	else
+		dsb = umem_off2ptr(umm, *dsb_off);
+
+	if (dsb == NULL)
+		D_GOTO(out, rc = 1);
+
+	D_ASSERT(dsb->dsb_magic == DTX_CMT_BLOB_MAGIC);
+
+	cont->vc_reindex_cmt_dtx = 1;
+
+	for (i = 0; i < dsb->dsb_count; i++) {
+		if (daos_is_zero_dti(&dsb->dsb_commmitted_data[i].dce_xid) ||
+		    dsb->dsb_commmitted_data[i].dce_epoch == 0) {
+			D_WARN("Skip invalid committed DTX entry\n");
+			continue;
+		}
+
+		D_ALLOC_PTR(dce);
+		if (dce == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+
+		D_INIT_LIST_HEAD(&dce->dce_bcb_link);
+		memcpy(&dce->dce_base, &dsb->dsb_commmitted_data[i],
+		       sizeof(dce->dce_base));
```

Comment:
or you could just dce->dce_base = dsb->dsb_commmitted_data[i];

Reply:
Which is more efficient?